### PR TITLE
Preserve local scopes in generator MIR

### PR DESCRIPTION
**Pull request by tmandry**
Part of #52924, depended upon by the generator layout optimization #60187.

This PR adds `StorageDead` statements in more places in generators, so we can see when non-`Drop` locals have gone out of scope and recover their storage.

The reason this is only done for generators is compiler performance. See https://github.com/rust-lang/rust/pull/60187#issuecomment-485637811 for what happens when we do this for all functions.

For `Drop` locals, we modify the `MaybeStorageLive` analysis to use `drop` to indicate that storage is no longer live for the local. Once `drop` returns or unwinds to our function, we implicitly assume that the local is `StorageDead`.

Instead of using `drop`, it is possible to emit more `StorageDead` statements in the MIR for `Drop` locals so we can handle all locals the same. I am fine with doing it that way, but this was the simplest approach for my purposes. It is also likely to be more performant.

r? @Zoxc (feel free to reassign)
cc @cramertj @eddyb @RalfJung @rust-lang/wg-async-await

### CODE REVIEW

**Comment by tmandry** [comment]
cc @rust-lang/lang 

The assumption this PR is making is that once `drop` returns to a function, whether it succeeds or panics, it is UB for the function to then access that local.

Let me know if I need to file an issue separately to discuss this!

File: src/librustc_mir/build/scope.rs [hunk 1]
```diff
@@ -138,21 +144,7 @@ struct DropData<'tcx> {
...
-    /// array)
-    unwind: Option<BasicBlock>,
-
-    /// The cached block for unwinds during cleanups-on-generator-drop path
-    ///
-    /// This is split from the standard unwind path here to prevent drop
-    /// elaboration from creating drop flags that would have to be captured
-    /// by the generator. I'm not sure how important this optimization is,
-    /// but it is here.
-    generator_drop: Option<BasicBlock>,
```
Comment (Zoxc) [inline src/librustc_mir/build/scope.rs]: Why is this removed?

Comment (tmandry) [reply src/librustc_mir/build/scope.rs]: The effect of this commit is to undo the optimization mentioned in the comment. We need the drop flags (`StorageDead` statements) that it removed, in order to do the analysis for the generator layout optimization.

Comment (tmandry) [reply src/librustc_mir/build/scope.rs]: So this unifies the `unwind` and `generator_drop` paths again.

**Comment by cramertj** [comment]
For additional context, @RalfJung pointed out that under stacked borrows, the creation of the `&mut Self` reference in `Drop::drop` means that all existing raw pointers to the local are invalidated, so the requirement introduced in this PR is likely already enforced under the existing model (though there was also some mention that `drop` is already kind of "special", so it might be that we could create some kind of exception here).

This means that code like this is UB: [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bb17a99761dc14aecb6aed29ff62c5bc)
Note that when run under MIRI, this code already produces stacked-borrows errors.

File: src/librustc_mir/build/scope.rs (hunk 1 above)
Comment (Zoxc) [reply src/librustc_mir/build/scope.rs]: > The effect of this commit is to undo the optimization mentioned in the comment. We need the drop flags (`StorageDead` statements) that it removed, in order to do the analysis for the generator layout optimization.

That doesn't make sense. `StorageDead` is unrelated to drop flags. We do not want drop flags to be created.
Additionally `unwind` has to drop generator arguments (we probably want to add those back) while the `generator_drop` path does not have them. I think you should just generate `StorageDead` statements for the `generator_drop` path, but not do the other changes.

Comment (tmandry) [reply src/librustc_mir/build/scope.rs]: Oh I see, yeah I completely misunderstood what was going on here.

Still, from looking at the code I couldn't find any places where the two paths are handled any differently (after my change to generate `StorageDead` statements for the `generator_drop` path, which is why I thought this change was okay. So, I guess I'm still not 100% sure what's going on here :)

Comment (tmandry) [reply src/librustc_mir/build/scope.rs]: I'll split up the commit so this is easier to see.

**Comment by Centril** [comment]
This seems reasonable to me; the only thing I would ask is that we document the assumed invariant in a less ephemeral place than the code-base since it seems easy to forget for users, lang designers, and compiler devs alike.

File: src/librustc_mir/build/scope.rs (hunk 1 above)
Comment (tmandry) [reply src/librustc_mir/build/scope.rs]: I split up the commit and reverted this change (but left it in the history for now).

**Comment by tmandry** [comment]
> the only thing I would ask is that we document the assumed invariant in a less ephemeral place than the code-base

There is an issue on the reference to document guarantees around drop and panicing: https://github.com/rust-lang-nursery/reference/issues/348

**Comment by Zoxc** [comment]
We probably want a pass to remove the StorageDead statements in the unwind path after the generator transformation.

**Comment by RalfJung** [comment]
> The reason this is only done for generators is compiler performance. See #60187 (comment) for what happens when we do this for all functions.

Do you think it would make sense to have a `-Z` flag for this that Miri could set, so that Miri checks these stricter rules in more places?

OTOH, this seemingly only affects the unwind path, of which test coverage is generally spotty at best. ;)  (And Miri doesn't even support unwinding yet.)

>  @RalfJung pointed out that under stacked borrows, the creation of the &mut Self reference in Drop::drop means that all existing raw pointers to the local are invalidated, so the requirement introduced in this PR is likely already enforced under the existing model.

I realized later that there is a gap in this argument: the `drop` implementation itself could create a raw pointer and leak it into global state, and taht could legitimately be used later.

Relying on `StorageDead` is probably better.

> Let me know if I need to file an issue separately to discuss this!

I don't think we have a topic yet to track `StorageDead`/`StorageLive` issues in general. That would also cover https://github.com/rust-lang/rust/issues/42371 then.

I'll open one in the UCG repository.

---

> For Drop locals, we modify the MaybeStorageLive analysis to use drop to indicate that storage is no longer live for the local. Once drop returns or unwinds to our function, we implicitly assume that the local is StorageDead.

Okay, so this is the meat of the PR as far as I am concerned: this changes the semantics of `Drop` to implicitly have a `StorageDead` in both the return and the unwind edge. If we decide to go with this approach, please open an issue in Miri so that we can implement this in Miri's semantics. It would also be helpful if you could come up with some example code that should be flagged as UB.

I have a few questions about this though:

* `Drop` works on an arbitrary place, `StorageDead` works on a local. I suppose the new semantics only occurs when the place is exactly a local, and does not involve any projections or so?

* In early passes of MIR construction (before drop flag elaboration), `Drop` is actually a conditional drop. Does the `StorageDead` only occur if the `Drop` happens, or does it always occur? Are we sure that drop flag elaboration is semantics-preserving under whatever choice we make here (and document, I hope)?

* What about `DropAndReplace`?

**Comment by Zoxc** [comment]
I think drop elaboration can turn a drop of a local into drops of its fields, which would effectively remove the implicit `StorageDead`.

**Comment by RalfJung** [comment]
Hm, that might be a problem. Depending on how we decide about the semantics for `StorageLive`, when `_3` is live doing `StorageDead(_3); StorageLive(_3)` might be okay while just `StorageLive(_3)` is not okay.

**Comment by tmandry** [comment]
> We probably want a pass to remove the StorageDead statements in the unwind path after the generator transformation.

We already remove StorageLive/StorageDead statements for any of the saved locals. For the rest of them, would it buy us anything to remove them?

**Comment by Zoxc** [comment]
@tmandry There's a list of reasons [here](https://github.com/rust-lang/rust/pull/60840/files#diff-efe92a4bb85e2e14ca66c492d897de28R105). Not sure how important it is.

**Comment by cramertj** [comment]
Would it be possible to split further discussion of the proper semantics of `StorageDead`/`StorageLive` out into a separate issue, and continue with landing this PR? I think there's general agreement that touching variables after `drop` has been called on them is undefined behavior, which AFAICT is the only guarantee necessary to make the optimization this unblocks valid.

Note that this PR is blocking a significant generator size optimization (https://github.com/rust-lang/rust/pull/60187) which is currently blocking the stabilization of async/await (https://github.com/rust-lang/rust/issues/52924). The lack of this optimization is currently causing frequent stack overflows and memory usage issues for production users (Fuchsia), as well as other nightly users of async/await.

I'd love it if we could land these PRs, unblocking async/await and fixing these problems for production users, and continue asynchronous discussion on the best way to model things in MIR (which, AFAICT, does not affect the validity of the optimization here). Is there a blocking issue that needs to be addressed first before we can merge?

**Comment by RalfJung** [comment]
> Would it be possible to split further discussion of the proper semantics of StorageDead/StorageLive out into a separate issue, and continue with landing this PR?

Fine for me. Feel free to open an issue. :)

The worst-case outcome I can see here is that we end up having to materialize the implicit `StorageDead` after `drop` in generators, to get the semantics straight. Would that be a deal-breaker?

**Comment by cramertj** [comment]
> Fine for me. Feel free to open an issue. :)

Opened https://github.com/rust-lang/rust/issues/61015.

> The worst-case outcome I can see here is that we end up having to materialize the implicit StorageDead after drop in generators, to get the semantics straight. Would that be a deal-breaker?

I certainly don't have any opposition to this in theory, but it seems worth experimenting to see what solutions we can come up with that don't significantly regress compiler performance (or other ways we can recover the compiler performance).

In the meantime, let's get this in--
@bors r+

**Comment by Zoxc** [comment]
@cramertj Did you actually review the PR?

**Comment by tmandry** [comment]
> @tmandry There's a list of reasons [here](https://github.com/rust-lang/rust/pull/60840/files#diff-efe92a4bb85e2e14ca66c492d897de28R105). Not sure how important it is.

Well, the idea of "observing a frozen stack" is probably incompatible with this change (and certainly incompatible with the optimization it enables for generators). Technically it isn't _wrong_ to remove the `StorageDead` after the optimization runs; it's equivalent to recovering these "frozen stack" semantics for the locals that didn't happen to be saved in the generator. But I don't think there's a good reason why we'd want that, and it could lead to more confusion.

The reasons given in the code aren't all that compelling because it boils down to "we thought clang did it this way but actually it doesn't." My understanding is that we never changed the behavior because that would hurt performance. However, we're already going to pay the cost here of emitting these `StorageDead`s in generators, and I doubt removing them later will recover that cost.

**Comment by cramertj** [comment]
@Zoxc Yes, I looked over it and it looked reasonable to me, but I'm much less familiar with this code than you are, so if you have specific concerns you'd like to see addressed before merging feel free to r-.

**Pull request merged**